computer_vision/object_classification/weight_converter.cpp

```diff
@@ -0,0 +1,103 @@
+/**
+ * @file object_classification.hpp
+ * @author Kartik Dutt
+ *
+ * Contains implementation of object classification suite. It can be used
+ * to select object classification model, it's parameter dataset and
+ * other training parameters.
+ *
+ * NOTE: This code needs to be adapted as this implementation doesn't support
+ *       Command Line Arguments.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#include <mlpack/core.hpp>
+#include <dataloader/dataloader.hpp>
+#include <models/darknet/darknet.hpp>
+#include <utils/utils.hpp>
+#include <ensmallen_utils/print_metric.hpp>
+#include <ensmallen_utils/periodic_save.hpp>
+#include <ensmallen.hpp>
+#include <mlpack/methods/ann/layer_names.hpp>
+
+using namespace mlpack;
+using namespace mlpack::ann;
+using namespace arma;
+using namespace std;
+using namespace ens;
+
+struct output : public boost::static_visitor<>
+{
+  template <class T>
+  void operator()(T t) const
+  {
+    // We could use this as a summary.
+    LayerNameVisitor a1;
+    t.Parameters();
+    std::cout << a1.LayerString(&t) << std::endl;
+    return;
+  }
+};
+
+/**
+ * std::cout << boost::get<Sequential<> *>(
+ *  darknetModel.GetModel().Model()[3])->Parameters().n_elem << std::endl;
+ */
+int main()
+{
+  #if defined(_OPENMP)
+    std::cout << "Compiled with OpenMP!" << std::endl;
+  #endif
+  DarkNet<> darknetModel(3, 224, 224, 1000, "none", true);
+  std::cout << "Model Compiled" << std::endl;
+
+  // std::cout << darknetModel.GetModel().Parameters().n_rows << " " <<
+  //    darknetModel.GetModel().Parameters().n_cols << std::endl;
+
+  size_t outSize = 32;
+  Convolution<> *layer = new Convolution<>(3, 32, 3, 3, 1, 1, 1, 1, 224, 224);
+  // std::cout << layer->Parameters().n_elem << std::endl;
+  size_t layerElement = layer->Parameters().n_elem - outSize;
+
+  // This will be equal to output size for conv layers.
+  // In darknet model bias term is removed.
+  size_t offset = 0;
+  size_t biasOffset = outSize;
+
+  arma::mat ConvWeights;
+  mlpack::data::Load("../../PyTorch-mlpack-DarkNet-Weight-Converter/conv_1_1.csv",
+      ConvWeights);
+  // Transpose weights to match FFN Class.
+  ConvWeights = ConvWeights.t();
+  std::cout << ConvWeights.n_rows << " " << ConvWeights.n_cols << std::endl;
+
+  darknetModel.GetModel().Parameters()(arma::span(offset, offset + layerElement - 1),
+      arma::span()) = ConvWeights;
+  
```

Comment:
I am able to load the weights from a csv into the FFN class. Here the [link](https://github.com/kartikdutt18/PyTorch-mlpack-DarkNet-Weight-Converter/blob/master/darknet.ipynb) to how the csv was generated. So I this is what we can do :
1. Create csv files for all layers that have weights.
2. Create a map with all layers and corresponding csv file path. Calculating offset and transferring weights to the FFN weights as shown above. 
I think this could work. To test it I guess I would have to download the imagenet dataset and run validation on it.
Kindly let me know what you think.
CC : @KimSangYeon-DGU, @saksham189, @zoq

Reply:
Then for I think we can avoid training models and simply transfer pytorch weights into mlpack and store them as ready to use binaries.

Reply:
Great, let's check the validation accuracy using the mlpack's Darknet-19 model :)
I think we can check the validation using [imagenette](https://github.com/fastai/imagenette) dataset that is a subset of imagenet

Reply:
Awesome, So I'll create a WeightConverter class to iterate over layers rather than manually adding weights. And we can test it out. I also found weights for DarkNet53.

Reply:
@kartikdutt18 Ahh, but the `imagenette` has only 10 classes, so I think we need to check the validation with `imagenet` dataset :D

Reply:
Since we are only generating prediction, I don't think that should be an issue but I can do imagenet as well. Kindly let me know what you think.

Reply:
About the validation set, there is a reference at [this](https://pjreddie.com/darknet/imagenet/) on the `Validating On ImageNet` part.

Reply:
> Since we are only generating prediction, I don't think that should be an issue but I can do imagenet as well. Kindly let me know what you think.

Ahh, makes sense :) 👍